src/Framework/AspNetCoreAnalyzers/src/Analyzers/RouteHandlers/DetectAmbiguousRoutes.cs

```diff
@@ -0,0 +1,215 @@
+            }
+
+            current = current.Parent;
+        }
+
+        return null;
+    }
+
+    /// <summary>
+    /// Test the invocation operation. Safe methods are those that we know don't add metadata that impacts metadata.
+    /// </summary>
+    private static bool IsAllowedEndpointBuilderMethod(IInvocationOperation invocationOperation, WellKnownTypes wellKnownTypes)
+    {
+        var method = invocationOperation.TargetMethod;
+
+        if (SymbolEqualityComparer.Default.Equals(method.ContainingType, wellKnownTypes.Get(WellKnownType.Microsoft_AspNetCore_Builder_RoutingEndpointConventionBuilderExtensions)))
+        {
+            return method.Name switch
+            {
+                "RequireHost" => false, // Adds IHostMetadata
+                "WithDisplayName" => true,
+                "WithMetadata" => false, // Can add anything
+                "WithName" => true,
+                "WithGroupName" => true,
+                _ => false
+            };
+        }
+        else if (SymbolEqualityComparer.Default.Equals(method.ContainingType, wellKnownTypes.Get(WellKnownType.Microsoft_AspNetCore_Builder_AuthorizationEndpointConventionBuilderExtensions)))
+        {
+            return method.Name switch
+            {
+                "RequireAuthorization" => true,
+                "AllowAnonymous" => true,
+                _ => false
+            };
+        }
+        else if (SymbolEqualityComparer.Default.Equals(method.ContainingType, wellKnownTypes.Get(WellKnownType.Microsoft_AspNetCore_Http_OpenApiRouteHandlerBuilderExtensions)))
+        {
+            return method.Name switch
+            {
+                "Accepts" => false, // Adds IAcceptsMetadata
+                "ExcludeFromDescription" => true,
+                "Produces" => true,
+                "ProducesProblem" => true,
+                "ProducesValidationProblem" => true,
+                "WithDescription" => true,
+                "WithSummary" => true,
+                "WithTags" => true,
+                _ => false
+            };
+        }
+        else if (SymbolEqualityComparer.Default.Equals(method.ContainingType, wellKnownTypes.Get(WellKnownType.Microsoft_AspNetCore_Builder_CorsEndpointConventionBuilderExtensions)))
+        {
+            return method.Name == "RequireCors";
+        }
+        else if (SymbolEqualityComparer.Default.Equals(method.ContainingType, wellKnownTypes.Get(WellKnownType.Microsoft_Extensions_DependencyInjection_OutputCacheConventionBuilderExtensions)))
+        {
+            return method.Name == "CacheOutput";
+        }
+        else if (SymbolEqualityComparer.Default.Equals(method.ContainingType, wellKnownTypes.Get(WellKnownType.Microsoft_AspNetCore_Builder_RateLimiterEndpointConventionBuilderExtensions)))
+        {
+            return method.Name switch
+            {
+                "RequireRateLimiting" => true,
+                "DisableRateLimiting" => true,
+                _ => false
+            };
```

Comment:
Stylistic suggestion

```suggestion
            return method.Name is "RequireRateLimiting" or "DisableRateLimiting";
```